src/sparse_matrix_builder.rs

```diff
@@ -0,0 +1,276 @@
+use std::collections::HashMap;
+use std::hash::BuildHasherDefault;
+use std::sync::Mutex;
+
+use crossbeam::queue::SegQueue;
+use dashmap::DashMap;
+use itertools::Itertools;
+use rayon::iter::IndexedParallelIterator;
+use rayon::iter::IntoParallelIterator;
+use rayon::iter::IntoParallelRefIterator;
+use rayon::iter::ParallelIterator;
+use rayon::prelude::ParallelSliceMut;
+use rustc_hash::FxHasher;
+
+use crate::entity::Hyperedge;
+use crate::sparse_matrix::{Edge, Entity, SparseMatrix, SparseMatrixDescriptor};
+
+#[derive(Debug, Default)]
+struct Row {
+    occurrence: u32,
+    row_sum: f32,
+}
+
+/// Data locality plays huge role in propagation phase performance
+/// We want connected nodes to have similar indices, as they will get updated together.
+/// NodeIndexer assigns successive indices to nodes connected via hyper-edges.
+/// Such ordering yields significant performance boost in propagation phase.
+#[derive(Debug)]
+pub struct NodeIndexer {
+    pub key_2_index: HashMap<u64, usize, BuildHasherDefault<FxHasher>>,
+    pub index_2_key: Vec<u64>,
+}
+
+/// Non-blocking builder suitable for multi-thread usage
+#[derive(Debug)]
+pub struct NodeIndexerBuilder {
+    col_a_id: u8,
+    col_b_id: u8,
+    key_2_index: DashMap<u64, usize, BuildHasherDefault<FxHasher>>,
+    index_2_key: Mutex<Vec<u64>>,
+    inserts_queue: SegQueue<u64>,
+}
+
+impl NodeIndexerBuilder {
+    pub fn new(descriptor: &SparseMatrixDescriptor) -> Self {
+        Self {
+            col_a_id: descriptor.col_a_id,
+            col_b_id: descriptor.col_b_id,
+            key_2_index: Default::default(),
+            index_2_key: Default::default(),
+            inserts_queue: Default::default(),
+        }
+    }
+
+    pub fn process(&self, hyperedge: &Hyperedge) {
+        for hash in hyperedge.nodes(self.col_a_id as usize) {
+            self.insert_nonblocking(&hash)
+        }
+        for hash in hyperedge.nodes(self.col_b_id as usize) {
+            self.insert_nonblocking(&hash)
+        }
+    }
+
+    pub fn finish(self) -> NodeIndexer {
+        let mut index_to_key = self
+            .index_2_key
+            .try_lock()
+            .expect("No one inserting at finish time")
+            .to_owned();
+        self.process_queue(&mut index_to_key);
+        NodeIndexer {
+            key_2_index: self.key_2_index.into_iter().collect(),
+            index_2_key: index_to_key,
+        }
+    }
+
+    fn insert_nonblocking(&self, key: &u64) {
+        if self.key_2_index.contains_key(key) {
+            return;
+        }
+        self.inserts_queue.push(*key);
+        if let Ok(mut index_2_key) = self.index_2_key.try_lock() {
```

Comment:
Wondering if you want to try_lock on each insert_nonblocking call? Maybe one should do it every few calls or randomly somehow?

Reply:
My assumption is that performance hit by checking mutex is negligible and it works fine as I test